File: 06-DevOps/05-IaC与配置管理/幂等模块/python/idem_modules.py
"""Ansible 幂等模块最小实现

来源:
- Ansible Official "playbooks_intro.html" (docs.ansible.com/ansible/devel/playbooks_intro.html)
  原文:"Most Ansible modules check whether the desired final state has
  already been achieved and exit without performing any actions if that
  state has been achieved. Repeating the task does not change the final
  state. Modules that behave this way are 'idempotent'."
- "Test Strategies" (docs.ansible.com/projects/ansible/latest/reference_appendices/test_strategies.html)
  "--check mode in Ansible can be used as a layer of testing as well.
  If running a deployment playbook against an existing system, using the
  --check flag to the ansible command will report if Ansible thinks it
  would have had to have made any changes to bring the system into a
  desired state"

实现 3 个核心 module 的幂等行为:
1. file     — 路径是否存在 + mode/owner 比对
2. package  — 是否已安装
3. service  — state 字段比对 (started/stopped/enabled)
"""

import json
import os
import grp  # POSIX
import pwd  # POSIX
from pathlib import Path
from typing import Any

# 模拟一个目标机事实:文件/包/服务
class Host:
    def __init__(self):
        self.fs = {  # path -> {mode, owner, group, content}
            "/etc/nginx/nginx.conf": {
                "mode": 0o644, "owner": "root", "group": "root",
                "content": "user www-data;\n",
            },
            "/var/log/app.log": {
                "mode": 0o600, "owner": "app", "group": "app",
                "content": "",
            },
            "/etc/missing.conf": None,
        }
        self.packages = {"nginx", "curl", "vim"}
        self.services = {  # name -> {running, enabled}
            "nginx": {"running": True,  "enabled": True},
            "sshd":  {"running": True,  "enabled": True},
            "redis": {"running": False, "enabled": False},
        }

    # ----- file ops -----
    def stat_file(self, path):
        if path not in self.fs:
            return None
        return self.fs[path]

    def write_file(self, path, *, mode, owner, group, content):
        self.fs[path] = {"mode": mode, "owner": owner, "group": group, "content": content}

    def remove_file(self, path):
        self.fs.pop(path, None)

    # ----- package -----
    def pkg_is_installed(self, name):
        return name in self.packages

    def pkg_install(self, name):
        self.packages.add(name)

    # ----- service -----
    def svc_state(self, name):
        return self.services.get(name)


# -----------------------------------------------------------------------------
# Module 1: file  (State=present/absent, mode/owner/content 校验)
# -----------------------------------------------------------------------------

def module_file(host: Host, params: dict, check_mode: bool = False) -> dict:
    """实现 ansible.builtin.file 风格:desired-state 检查

    返回 Ansible 经典 result:
        changed   — bool,本次是否真发生修改
        failed   — bool
        msg      — str
        diff     — dict('before', 'after'),仅 --diff 时填
    """
    path = params["path"]
    state = params.get("state", "file")  # file / directory / link / absent / touch
    desired_mode = params.get("mode")
    desired_owner = params.get("owner")
    desired_group = params.get("group")
    desired_content = params.get("_content")  # demo 用

    cur = host.stat_file(path)
    result = {"changed": False, "failed": False, "msg": "", "diff": {}}

    if state == "absent":
        if cur is not None:
            result["changed"] = True
            result["diff"] = {"before": dict(cur), "after": None}
            if not check_mode:
                host.remove_file(path)
            result["msg"] = "file removed"
        else:
            result["msg"] = "file already absent"
        return result

    # state = file / touch / directory (本 demo 仅 file)
    if cur is None:
        # 全新创建
        result["changed"] = True
        new = {"mode": desired_mode, "owner": desired_owner,
               "group": desired_group, "content": desired_content or ""}
        result["diff"] = {"before": None, "after": new}
        if not check_mode:
            host.write_file(path, **new)
        result["msg"] = "file created"
        return result

    # 已存在: 逐字段比对, 任一差异 → changed=True
    diffs = {}
    if desired_mode is not None and cur["mode"] != desired_mode:
        diffs["mode"] = {"before": cur["mode"], "after": desired_mode}
    if desired_owner is not None and cur["owner"] != desired_owner:
        diffs["owner"] = {"before": cur["owner"], "after": desired_owner}
    if desired_group is not None and cur["group"] != desired_group:
        diffs["group"] = {"before": cur["group"], "after": desired_group}
    if desired_content is not None and cur["content"] != desired_content:
        diffs["content"] = {"before": cur["content"], "after": desired_content}
    if diffs:
        result["changed"] = True
        result["diff"] = {"before": {k: v["before"] for k, v in diffs.items()},
                           "after":  {k: v["after"]  for k, v in diffs.items()}}
        if not check_mode:
            new = dict(cur)
            new.update({k: v["after"] for k, v in diffs.items()})
            host.write_file(path, **new)
        result["msg"] = "file updated"
    else:
        result["msg"] = "file already in desired state"
    return result


# -----------------------------------------------------------------------------
# Module 2: package (State=present/absent, name=...)
# -----------------------------------------------------------------------------

def module_package(host: Host, params: dict, check_mode: bool = False) -> dict:
    name = params["name"]
    state = params.get("state", "present")
    result = {"changed": False, "failed": False, "msg": "", "diff": {}}
    installed = host.pkg_is_installed(name)

    if state == "present":
        if installed:
            result["msg"] = f"package {name} already installed"
        else:
            result["changed"] = True
            result["diff"] = {"before": "absent", "after": "present"}
            if not check_mode:
                host.pkg_install(name)
            result["msg"] = f"installed package {name}"
    elif state == "absent":
        if not installed:
            result["msg"] = f"package {name} already absent"
        else:
            result["changed"] = True
            result["diff"] = {"before": "present", "after": "absent"}
            result["msg"] = f"removed package {name}"
    return result


# -----------------------------------------------------------------------------
# Module 3: service (State=started/stopped, enabled=true/false)
# -----------------------------------------------------------------------------

def module_service(host: Host, params: dict, check_mode: bool = False) -> dict:
    name = params["name"]
    desired_running = params.get("state") in ("started", "running")
    desired_enabled = params.get("enabled")
    cur = host.svc_state(name)
    result = {"changed": False, "failed": False, "msg": "", "diff": {}}

    if cur is None:
        result["failed"] = True
        result["msg"] = f"service {name} does not exist on host"
        return result

    diff = {}
    if cur["running"] != desired_running:
        diff["state"] = {"before": cur["running"], "after": desired_running}
    if desired_enabled is not None and cur["enabled"] != desired_enabled:
        diff["enabled"] = {"before": cur["enabled"], "after": desired_enabled}
    if diff:
        result["changed"] = True
        result["diff"] = {"before": {k: v["before"] for k, v in diff.items()},
                          "after":  {k: v["after"]  for k, v in diff.items()}}
        if not check_mode:
            if "state" in diff:
                cur["running"] = desired_running
            if "enabled" in diff:
                cur["enabled"] = desired_enabled
        result["msg"] = f"service {name} updated"
    else:
        result["msg"] = f"service {name} already in desired state"
    return result


# -----------------------------------------------------------------------------
# Demo playbook runner: 同一 playbook 跑三遍,观察 idempotent 是否成立
# -----------------------------------------------------------------------------

def run_playbook(host):
    """play = 一组 module 调用,执行后打印 changed 计数"""
    print("--- Playbook: 3 module 期望收敛到 desired state ---")
    results = []
    results.append(("file: nginx.conf mode=0o640 owner=root",
                    module_file(host, {"path": "/etc/nginx/nginx.conf",
                                       "mode": 0o640, "owner": "root",
                                       "group": "root"})))
    results.append(("file: new file /tmp/x.txt 0o755 content='hello'",
                    module_file(host, {"path": "/tmp/x.txt", "mode": 0o755,
                                       "owner": "root", "group": "root",
                                       "_content": "hello"})))
    results.append(("file: /etc/missing.conf absent",
                    module_file(host, {"path": "/etc/missing.conf",
                                       "state": "absent"})))
    results.append(("package: jq installed",
                    module_package(host, {"name": "jq", "state": "present"})))
    results.append(("service: redis started+enabled",
                    module_service(host, {"name": "redis",
                                          "state": "started",
                                          "enabled": True})))
    results.append(("service: nginx started (already)",
                    module_service(host, {"name": "nginx",
                                          "state": "started",
                                          "enabled": True})))
    for label, r in results:
        flag = "CHANGED" if r["changed"] else "ok"
        print(f"  [{flag:7}] {label:50s}  msg='{r['msg']}'")
    changed = sum(1 for _, r in results if r["changed"])
    print(f"  → 累计 changed = {changed}/{len(results)}\n")
    return results


def demo_idempotency():
    """同一 playbook 跑 3 次,只有第 1 次有 changed"""
    print("=== Ansible 幂等模块 demo ===\n")
    host = Host()
    print("--- 第 1 次执行:期望全部触发 ---")
    r1 = run_playbook(host)
    print("--- 第 2 次执行:幂等性 — 期望全部 ok ---")
    r2 = run_playbook(host)
    print("--- 第 3 次执行:继续幂等 — 全部 ok ---")
    r3 = run_playbook(host)

    # 结论
    c1 = sum(1 for _, r in r1 if r["changed"])
    c2 = sum(1 for _, r in r2 if r["changed"])
    c3 = sum(1 for _, r in r3 if r["changed"])
    print(f"--- 结论 ---")
    print(f"  1st pass: {c1}/6 changed   (期望 ≥ 1)")
    print(f"  2nd pass: {c2}/6 changed   (期望 0,证明幂等)")
    print(f"  3rd pass: {c3}/6 changed   (期望 0,继续幂等)")
    assert c1 >= 1 and c2 == 0 and c3 == 0, "idempotency 失败"
    print("  ✓ idempotency 验证通过\n")


def demo_check_mode():
    """--check:不真改 host,只报 diff"""
    print("--- Demo: --check mode 不真改 ---")
    host = Host()
    pre_nginx = host.stat_file("/etc/nginx/nginx.conf")
    r = module_file(host, {"path": "/etc/nginx/nginx.conf",
                           "mode": 0o600, "owner": "root"}, check_mode=True)
    post_nginx = host.stat_file("/etc/nginx/nginx.conf")
    print(f"  check_mode=True → changed={r['changed']}, msg='{r['msg']}'")
    print(f"  diff (would-be): {json.dumps(r['diff'], default=str)}")
    print(f"  host 真实状态保持一致? {pre_nginx == post_nginx}")
    assert pre_nginx == post_nginx, "--check must not modify"
    print("  ✓ check_mode 未真实修改\n")


def main():
    demo_idempotency()
    demo_check_mode()


if __name__ == "__main__":
    main()

File: 06-DevOps/05-IaC与配置管理/幂等模块/go/idem_modules.go
// Ansible 幂等模块最小 Go 镜像
//
// 来源:Ansible 官方 docs.ansible.com/ansible/devel/playbooks_intro.html
//  "Modules that behave this way are 'idempotent'."
//  "Repeating the task does not change the final state"
//
// Go 版实现 file/package/service 三个最小幂等模块,验证连跑 N 次 changed 数下降

package main

import "fmt"

// -----------------------------------------------------------------------------
// Host — 模拟文件系统 + 包 + 服务
// -----------------------------------------------------------------------------

type FileInfo struct {
	Mode    uint32
	Owner   string
	Group   string
	Content string
}

type ServiceState struct {
	Running bool
	Enabled bool
}

type Host struct {
	Files    map[string]*FileInfo
	Packages map[string]bool
	Services map[string]*ServiceState
}

func NewHost() *Host {
	return &Host{
		Files: map[string]*FileInfo{
			"/etc/nginx/nginx.conf": {Mode: 0644, Owner: "root", Group: "root", Content: "user www-data;\n"},
			"/var/log/app.log":      {Mode: 0600, Owner: "app", Group: "app", Content: ""},
		},
		Packages: map[string]bool{"nginx": true, "curl": true, "vim": true},
		Services: map[string]*ServiceState{
			"nginx": {Running: true, Enabled: true},
			"sshd":  {Running: true, Enabled: true},
			"redis": {Running: false, Enabled: false},
		},
	}
}

// -----------------------------------------------------------------------------
// Module result — 对应 Ansible result dict
// -----------------------------------------------------------------------------

type ModuleResult struct {
	Changed bool                   `json:"changed"`
	Failed  bool                   `json:"failed"`
	Msg     string                 `json:"msg"`
	Diff    map[string]interface{} `json:"diff"`
}

// -----------------------------------------------------------------------------
// Module: file
// -----------------------------------------------------------------------------

func ModuleFile(h *Host, params map[string]interface{}, checkMode bool) ModuleResult {
	path := params["path"].(string)
	state, _ := params["state"].(string) // file/absent
	desiredMode, _ := params["mode"].(uint32)
	desiredOwner, _ := params["owner"].(string)
	desiredGroup, _ := params["group"].(string)
	desiredContent, _ := params["_content"].(string)

	result := ModuleResult{Diff: map[string]interface{}{}}

	if state == "absent" {
		if _, exists := h.Files[path]; exists {
			result.Changed = true
			result.Diff = map[string]interface{}{"before": "exists", "after": nil}
			if !checkMode {
				delete(h.Files, path)
			}
			result.Msg = "file removed"
		} else {
			result.Msg = "file already absent"
		}
		return result
	}

	cur, exists := h.Files[path]
	if !exists {
		// new file
		newF := &FileInfo{Mode: desiredMode, Owner: desiredOwner, Group: desiredGroup, Content: desiredContent}
		result.Changed = true
		result.Diff = map[string]interface{}{"before": nil, "after": newF}
		if !checkMode {
			h.Files[path] = newF
		}
		result.Msg = "file created"
		return result
	}

	// existing: per-field diff
	diffs := map[string]interface{}{}
	if cur.Mode != desiredMode && desiredMode != 0 {
		diffs["mode"] = map[string]interface{}{"before": cur.Mode, "after": desiredMode}
	}
	if cur.Owner != desiredOwner && desiredOwner != "" {
		diffs["owner"] = map[string]interface{}{"before": cur.Owner, "after": desiredOwner}
	}
	if cur.Group != desiredGroup && desiredGroup != "" {
		diffs["group"] = map[string]interface{}{"before": cur.Group, "after": desiredGroup}
	}
	if cur.Content != desiredContent && desiredContent != "" {
		diffs["content"] = map[string]interface{}{"before": cur.Content, "after": desiredContent}
	}
	if len(diffs) > 0 {
		result.Changed = true
		result.Diff = map[string]interface{}{"before": diffKeys(diffs, "before"), "after": diffKeys(diffs, "after")}
		if !checkMode {
			for k, v := range diffs {
				m := v.(map[string]interface{})
				switch k {
				case "mode":
					cur.Mode = m["after"].(uint32)
				case "owner":
					cur.Owner = m["after"].(string)
				case "group":
					cur.Group = m["after"].(string)
				case "content":
					cur.Content = m["after"].(string)
				}
			}
		}
		result.Msg = "file updated"
	} else {
		result.Msg = "file already in desired state"
	}
	return result
}

// diffKeys: 把 "diff.before.X" 提取为一个新 map
func diffKeys(d map[string]interface{}, which string) map[string]interface{} {
	out := map[string]interface{}{}
	for k, v := range d {
		if m, ok := v.(map[string]interface{}); ok { out[k] = m[which] }
	}
	return out
}

// -----------------------------------------------------------------------------
// Module: package
// -----------------------------------------------------------------------------

func ModulePackage(h *Host, params map[string]interface{}, _ bool) ModuleResult {
	name := params["name"].(string)
	state, _ := params["state"].(string)
	result := ModuleResult{Diff: map[string]interface{}{}}
	installed, exists := h.Packages[name]
	if state == "present" {
		if exists && installed {
			result.Msg = fmt.Sprintf("package %s already installed", name)
		} else {
			result.Changed = true
			result.Diff = map[string]interface{}{"before": "absent", "after": "present"}
			h.Packages[name] = true
			result.Msg = fmt.Sprintf("installed package %s", name)
		}
	} else if state == "absent" {
		if !exists || !installed {
			result.Msg = fmt.Sprintf("package %s already absent", name)
		} else {
			result.Changed = true
			result.Diff = map[string]interface{}{"before": "present", "after": "absent"}
			delete(h.Packages, name)
			result.Msg = fmt.Sprintf("removed package %s", name)
		}
	}
	return result
}

// -----------------------------------------------------------------------------
// Module: service
// -----------------------------------------------------------------------------

func ModuleService(h *Host, params map[string]interface{}, checkMode bool) ModuleResult {
	name := params["name"].(string)
	desiredState, _ := params["state"].(string) // started/stopped
	desiredEnabled, hasEnabled := params["enabled"].(bool)

	result := ModuleResult{Diff: map[string]interface{}{}}
	cur, exists := h.Services[name]
	if !exists {
		result.Failed = true
		result.Msg = fmt.Sprintf("service %s does not exist", name)
		return result
	}
	diffs := map[string]interface{}{}
	desiredRunning := desiredState == "started" || desiredState == "running"
	if cur.Running != desiredRunning {
		diffs["state"] = map[string]interface{}{"before": cur.Running, "after": desiredRunning}
	}
	if hasEnabled && cur.Enabled != desiredEnabled {
		diffs["enabled"] = map[string]interface{}{"before": cur.Enabled, "after": desiredEnabled}
	}
	if len(diffs) > 0 {
		result.Changed = true
		result.Diff = map[string]interface{}{"before": diffKeys(diffs, "before"), "after": diffKeys(diffs, "after")}
		if !checkMode {
			for k, v := range diffs {
				m := v.(map[string]interface{})
				switch k {
				case "state":
					cur.Running = m["after"].(bool)
				case "enabled":
					cur.Enabled = m["after"].(bool)
				}
			}
		}
		result.Msg = fmt.Sprintf("service %s updated", name)
	} else {
		result.Msg = fmt.Sprintf("service %s already in desired state", name)
	}
	return result
}

// -----------------------------------------------------------------------------
// Run a playbook once
// -----------------------------------------------------------------------------

type ModuleCall struct {
	Label  string
	Result ModuleResult
}

func RunPlaybook(h *Host) []ModuleCall {
	calls := []ModuleCall{}
	calls = append(calls, ModuleCall{"file: nginx.conf mode=0o640",
		ModuleFile(h, map[string]interface{}{
			"path": "/etc/nginx/nginx.conf", "mode": uint32(0640),
			"owner": "root", "group": "root"}, false)})
	calls = append(calls, ModuleCall{"file: /tmp/x.txt new content='hello'",
		ModuleFile(h, map[string]interface{}{
			"path": "/tmp/x.txt", "mode": uint32(0755),
			"owner": "root", "group": "root",
			"_content": "hello"}, false)})
	calls = append(calls, ModuleCall{"file: /etc/missing.conf absent",
		ModuleFile(h, map[string]interface{}{
			"path": "/etc/missing.conf", "state": "absent"}, false)})
	calls = append(calls, ModuleCall{"package: jq installed",
		ModulePackage(h, map[string]interface{}{
			"name": "jq", "state": "present"}, false)})
	calls = append(calls, ModuleCall{"service: redis started+enabled",
		ModuleService(h, map[string]interface{}{
			"name": "redis", "state": "started", "enabled": true}, false)})
	calls = append(calls, ModuleCall{"service: nginx started (already)",
		ModuleService(h, map[string]interface{}{
			"name": "nginx", "state": "started", "enabled": true}, false)})
	return calls
}

func printPlaybook(label string, calls []ModuleCall) {
	fmt.Printf("--- %s ---\n", label)
	changed := 0
	for _, c := range calls {
		flag := "ok"
		if c.Result.Changed {
			flag = "CHANGED"
			changed++
		}
		fmt.Printf("  [%7s] %-50s  msg=%q\n", flag, c.Label, c.Result.Msg)
	}
	fmt.Printf("  → changed = %d/%d\n\n", changed, len(calls))
}

func main() {
	fmt.Println("=== Ansible 幂等模块 demo (Go 版) ===\n")
	h := NewHost()
	printPlaybook("第 1 次执行:期望全部触发", RunPlaybook(h))
	r2 := RunPlaybook(h)
	printPlaybook("第 2 次执行:期望 0 changed", r2)
	r3 := RunPlaybook(h)
	printPlaybook("第 3 次执行:继续幂等", r3)

	c2, c3 := 0, 0
	for _, c := range r2 { if c.Result.Changed { c2++ } }
	for _, c := range r3 { if c.Result.Changed { c3++ } }
	fmt.Println("--- 结论 ---")
	fmt.Printf("  2nd pass changed = %d (期望 0)\n", c2)
	fmt.Printf("  3rd pass changed = %d (期望 0)\n", c3)
	if c2 == 0 && c3 == 0 { fmt.Println("  ✓ idempotency 验证通过") }
}
